feat: RBAC Documentation Drafting (#332)
input-output-hk/catalyst-voices: docs/src/catalyst-standards/draft-cips/x509-envelope-metadata/cip-xxxx.md

License: CC-BY-4.0
---

## Abstract

x509 Certificate metadata and related metadata within an x509 Registration/Update transaction must be
protected against replayability.  

## Motivation: why is this CIP necessary?

x509 certificate registration necessarily references data within the transaction body that it is attached to.
Preventing replayability of the verbatim Auxiliary data attached to a transaction prevents the relationships
between those references being altered and closes a potential attack vector where an unrelated registration update
is attached to a different transaction.

Rather than define an explicit issue with potential replayability of Metadata, we simply seek to prevent it
to close off a whole category of potential attacks.

## Specification

The [x509 Metadata Envelope CDDL] defines the technical specification of the Metadata Envelope.
The following description provides more information and context to that formal specification.
Where this description and that specification are in disagreement, the specification prevails.

The metadata envelope has the following basic structure:

![x509 Secure Metadata Envelope structure](images/metadata-envelope-structure.svg)

> The purpose of this data is to lock a particular set of Auxiliary data verifiably to its original transaction.

Reviewer: Maybe adding some details here would give more context to readers before the explanation made below.
Something like
```suggestion
The purpose of this data is to lock a particular set of Auxiliary data verifiably to its original transaction by using transactions input hash.
```

> It can not be permitted to take teh Auxiliary data to another transaction and just attach it without it being detectable.

Reviewer: ```suggestion
It can not be permitted to take the Auxiliary data to another transaction and just attach it without it being detectable.
```

| --- | --- |
| ca7a1457-ef9f-4c7f-9c74-7f8c4a4cfa6c | Project Catalyst User Role Registrations |
| ca7ad312-a19b-4412-ad53-2a36fb14e2e5 | Project Catalyst Admin Role Registrations |
  
The reason in this case for having two purposes for the same application is to allow the rules
governing registration to be easier managed between the groups.
It also allows Admins to have a distinct identity when they are just users of the system.
dApps are free top have as many "purposes" defined as suits their goals.

#### Key 1: txn-inputs-hash

This is a hash of the transaction inputs field from the transaction this auxiliary data was original attached to.
This hash anchors the auxiliary data to a particular transaction, if the auxiliary data is moved to a new transaction
it is impossible for the txn-inputs-hash to remain the same, as UTXO's can only be spent once.

This is a key to preventing replayability of the metadata, but is not enough by itself as it needs to be able to be
made immutable, such that any change to this field can be detected.

#### Key 2: previous_transaction_id

This is a 32 byte hash of the previous transaction in a chain of registration updates made by the same user.

> The only registration which may not have it for a single user is their first.

Reviewer: ```suggestion
The only registration which may not have it for a single user is when they first registered.
```
